firedrake/slate/preconditioners.py

```diff
@@ -0,0 +1,285 @@
+
+        # Set up the functions for the original, hybridized
+        # and schur complement systems
+        self.broken_solution = Function(V_d)
+        self.broken_rhs = Function(V_d)
+        self.trace_solution = Function(TraceSpace)
+        self.unbroken_solution = Function(V)
+        self.unbroken_rhs = Function(V)
+
+        # Create the symbolic Schur-reduction
+        Atilde = Tensor(new_form)
+        gammar = TestFunction(TraceSpace)
+        n = FacetNormal(V.mesh())
+
+        # Vector trial function will have a non-empty ufl_shape
+        sigma = next(f for f in TrialFunctions(V_d) if bool(f.ufl_shape))
+
+        # NOTE: Once extruded is ready, this will change slightly
+        # to include both horizontal and vertical interior facets
+        K = Tensor(gammar('+') * ufl.dot(sigma, n) * ufl.dS)
+
+        # Assemble the Schur complement operator and right-hand side
+        self.schur_rhs = Function(TraceSpace)
+        self._assemble_Srhs = create_assembly_callable(
+            K * Atilde.inv * self.broken_rhs,
+            tensor=self.schur_rhs,
+            form_compiler_parameters=context.fc_params)
+
+        schur_comp = K * Atilde.inv * K.T
+        self.S = allocate_matrix(schur_comp,
+                                 bcs=trace_conditions,
+                                 form_compiler_parameters=context.fc_params)
+        self._assemble_S = create_assembly_callable(
+            schur_comp,
+            tensor=self.S,
+            bcs=trace_conditions,
+            form_compiler_parameters=context.fc_params)
+
+        self._assemble_S()
+        self.S.force_evaluation()
+        Smat = self.S.petscmat
+
+        # Nullspace for the multiplier problem
+        nullsp = P.getNullSpace()
+        if nullsp.handle != 0:
+            new_vecs = get_trace_nullspace_vecs(K * Atilde.inv, nullsp,
+                                                V, V_d, TraceSpace)
+            tr_nullsp = PETSc.NullSpace().create(vectors=new_vecs,
+                                                 comm=pc.comm)
+            Smat.setNullSpace(tr_nullsp)
+
+        # Set up the KSP for the system of Lagrange multipliers
+        ksp = PETSc.KSP().create(comm=pc.comm)
+        ksp.setOptionsPrefix(prefix + "trace_")
+        ksp.setTolerances(rtol=1e-13)
+        ksp.setOperators(Smat)
+        ksp.setUp()
+        ksp.setFromOptions()
+        self.ksp = ksp
+
+        # Now we construct the local tensors for the reconstruction stage
+        # TODO: Add support for mixed tensors and these variables
+        # become unnecessary
+        split_forms = split_form(new_form)
+        A = Tensor(next(sf.form for sf in split_forms
+                        if sf.indices == (0, 0)))
+        B = Tensor(next(sf.form for sf in split_forms
+                        if sf.indices == (1, 0)))
+        C = Tensor(next(sf.form for sf in split_forms
+                        if sf.indices == (1, 1)))
```

Comment:
As a remark: when I recover using all the blocks `A, B, C, D` I lose some precision when measuring the error in `test_slate_hybridization.py`. Instead of `1e-11` relative error between hybridized and non-hybridized solutions, I'm hovering at around `1e-7`. I've tried changing the solver parameters, but opted to stay with what's currently there.

After running some resolution tests (not tests in this PR but on my machine), I still get expected convergence and accuracy results. I think this is something worth discussing at some point.